thali/thalireplicationmanager.js

```diff
@@ -38,25 +47,54 @@ ThaliReplicationManager.events = {
 };
 
 /**
-* Starts the Thali replication manager with the given device name and port number
-* @param {String} deviceName the device name to advertise.
+* Returns the current device's id if available. If not, the provided callback
+* function is saved and called when the id does become available.
+* @param {Function} cb the callback which returns the current device's id.
+*/
+ThaliReplicationManager.prototype.getDeviceIdentity = function (cb) {
+  if(this._deviceIdentityFlag == deviceIdentityFlag.deviceIdentityAvailable) {
+    cb(null, this._deviceName);
+  } else if(this._deviceIdentityFlag == deviceIdentityFlag.noDeviceIdentitySet) {
+    var cryptomanager = require('./thalicryptomanager');
+    this._deviceIdentityFlag = deviceIdentityFlag.gettingDeviceIdentity;
+    cryptomanager.getPublicKeyHash(function (err, publicKeyHash) {
+      if (err) {
+        this._deviceIdentityFlag = deviceIdentityFlag.noDeviceIdentitySet;
+        informDeviceIdentityListeners.call(this, err, null);
+        clearDeviceIdentityListeners.call(this);
+        cb(err);
+      } else {
+        this._deviceName = publicKeyHash;
+        this._deviceIdentityFlag = deviceIdentityFlag.deviceIdentityAvailable;
+        informDeviceIdentityListeners.call(this, null, publicKeyHash);
+        clearDeviceIdentityListeners.call(this);
+        cb(null, publicKeyHash);
+      }
+    }.bind(this));
+  } else {
+    // save the callback for future
+    this._deviceIdentityListeners.push(cb);
+  }
+}
+
+/**
+* Starts the Thali replication manager with the given port number and db name.
+* The device-id is obtained using the cryptomanager's API.
 * @param {Number} port the port number used for synchronization.
+* @param {String} dbName the db name to sync.
 */
-ThaliReplicationManager.prototype.start = function (deviceName, port, dbName) {
-  validations.ensureNonNullOrEmptyString(deviceName, 'deviceName');
+ThaliReplicationManager.prototype.start = function (port, dbName) {
   validations.ensureValidPort(port);
   validations.ensureNonNullOrEmptyString(dbName, 'dbName');
-
   this.emit(ThaliReplicationManager.events.STARTING);
-
+  
   this._port = port;
-  this._deviceName = deviceName;
   this._dbName = dbName;
   this._serverBridge = muxServerBridge.call(this, port);
   this._serverBridge.listen(function () {
     this._serverBridgePort = this._serverBridge.address().port;
 
-    this._emitter.startBroadcasting(deviceName, this._serverBridgePort, function (err) {
+    this._emitter.startBroadcasting(this._deviceName, this._serverBridgePort, function (err) {
```

Comment:
I apologize if I'm missing something but I don't see a check anywhere here that this._deviceName is actually set. There seems to be an assumption that getDeviceIdentity will always be called before start. That seems like a dangerous assumption to make. If in fact we are making that assumption then I would suggest that code should be 'fail safe'. That is, either start should fail if getDeviceIdentity hasn't been called or we should build getDeviceIdentity into start just to make sure. The later seems preferable.

Sorry if I missed where deviceName was guaranteed to be set.